File: src/mandelbrot.rs
use std::env;

fn mandelbrot_iterate(max_modulus: u32, max_iterations: u32, cr: f64, ci: f64) -> u32 {
    let mut zr = 0.0;
    let mut zi = 0.0;
    let mut rr;
    let mut ri;
    let mut rc;
    let mut i = 0;

    loop {
        rr = zr * zr;
        ri = zi * zi;
        rc = zr * zi;
        zr = rr - ri + cr;
        zi = 2.0 * rc + ci;
        i += 1;

        if (rr + ri) as u32 >= max_modulus {
            return i;
        }

        if i > max_iterations {
            return 0;
        }
    }
}

fn mandelbrot_draw(max_modulus: u32, max_iterations: u32, resolution: u32) {
    let unit = (resolution / 6) as u32;
    let width = unit * 3 * 2;
    let height = unit * 2;

    let mut cr: f64;
    let mut ci: f64;
    let mut result: u32;

    for j in 0..height + 1 {
        for i in 0..width + 1 {
            cr = i as f64 / unit as f64 / 2.0 - 2.0;
            ci = 1.0 - (j as f64 / unit as f64);

            result = mandelbrot_iterate(max_modulus, max_iterations, cr, ci) % 26;

            if result == 0 {
                print!(" ");
            } else {
                print!("{}", char::from_u32(64 + result).unwrap());
            }
        }

        print!("\n");
    }
}

fn main() {
    let max_modulus = match env::var_os("MANDELBROT_MAX_MODULUS") {
        Some(v) => v.into_string().unwrap(),
        None => panic!("$MANDELBROT_MAX_MODULUS not set.")
    };

    let max_iterations = match env::var_os("MANDELBROT_MAX_ITERATIONS") {
        Some(v) => v.into_string().unwrap(),
        None => panic!("$MANDELBROT_MAX_ITERATIONS not set.")
    };

    let resolution = match env::var_os("MANDELBROT_RESOLUTION") {
        Some(v) => v.into_string().unwrap(),
        None => panic!("$MANDELBROT_RESOLUTION not set.")
    };

    mandelbrot_draw(
        max_modulus.parse::<u32>().unwrap(),
        max_iterations.parse::<u32>().unwrap(),
        resolution.parse::<u32>().unwrap()
    );

    println!("Max modulus: {}", max_modulus);
    println!("Max iterations: {}", max_iterations);
    println!("Resolution: {}", resolution);
}

File: src/mandelbrot.pas
program mandelbrot;

uses SysUtils;

function MandelbrotIterate(MaxModulus, MaxIterations: Cardinal; cr, ci: Real): Cardinal;
var
    i: Cardinal;
    zr, zi, rr, ri, rc: Real;
begin
    zr := 0.0;
    zi := 0.0;
    i := 0;

    while True do
    begin
        rr := zr * zr;
        ri := zi * zi;
        rc := zr * zi;

        zr := rr - ri + cr;
        zi := 2 * rc + ci;

        Inc(i);

        if (rr + ri) > MaxModulus then
        begin
            MandelbrotIterate := i;
            Break;
        end;

        if i > MaxIterations then
        begin
            MandelbrotIterate := 0;
            Break;
        end;
    end;
end;

procedure MandelbrotDraw(MaxModulus, MaxIterations, Resolution: Cardinal);
var
    cr, ci: Real;
    unt, rst, width, height: Cardinal;
    i, j: Cardinal;
begin
    unt := resolution div 6;
    width := unt * 3 * 2;
    height := unt * 2;

    for j := 0 to height do
    begin
        for i := 0 to width do
        begin
            cr := (i / unt ) / 2.0 - 2.0;
            ci := 1.0 - j / unt;

            rst := MandelbrotIterate(MaxModulus, MaxIterations, cr, ci) mod 26;

            if rst = 0 then
                Write(' ')
            else
                Write(chr(64 + rst));
        end;

        WriteLn();
    end;

end;

procedure Main();
var
    MaxModulus, MaxIterations, Resolution: Cardinal;
begin
    MaxModulus := StrToInt(GetEnvironmentVariable('MANDELBROT_MAX_MODULUS'));
    MaxIterations := StrToInt(GetEnvironmentVariable('MANDELBROT_MAX_ITERATIONS'));
    Resolution := StrToInt(GetEnvironmentVariable('MANDELBROT_RESOLUTION'));

    MandelbrotDraw(MaxModulus, MaxIterations, Resolution);

    WriteLn(Format('Max modulus: %d', [MaxModulus]));
    WriteLn(Format('Max iterations: %d', [MaxIterations]));
    WriteLn(Format('Resolution: %d', [Resolution]));
end;

begin
    Main();
end.
